notebooks/02-ingestion/02-batch-ingestion.ipynb

```diff
@@ -0,0 +1,1008 @@
+    "\n",
+    "```\n",
+    "WITH \"ext\" AS \n",
+    "(\n",
+    "    SELECT ...\n",
+    "      FROM TABLE (EXTERN ( ... ) )\n",
+    ") EXTEND (...)\n",
+    "```\n",
+    "\n",
+    "`EXTERN` supports many batch [input sources](https://druid.apache.org/docs/latest/ingestion/native-batch-input-sources.html) and [formats](https://druid.apache.org/docs/latest/ingestion/data-formats.html). In this case we are using input source type `http` to access a set or `uris` that each contain a data file in the `json` data format. Note that compressed files are allowed and will automatically be decompressed.\n",
+    "```\n",
+    "FROM TABLE(\n",
+    "      EXTERN(\n",
+    "        '{\"type\":\"http\",\"uris\":[\"https://druid.apache.org/data/wikipedia.json.gz\"]}',\n",
+    "        '{\"type\":\"json\"}'\n",
+    "      )\n",
+    "```\n",
+    "The `EXTEND` clause describes the input schema using SQL data types:\n",
+    "```\n",
+    "EXTEND (\"isRobot\" VARCHAR, \"channel\" VARCHAR, \"timestamp\" VARCHAR,  ...)\n",
+    "```\n",
+    "\n",
+    "The final SELECT statement defines the transformations and schema of the resulting Druid table. A `__time` column is usually parsed from the source, this expression will be mapped to Druid's primary time partitioning of segments. In this case we specified the `__time` column and ingested the rest of the columns \"AS IS\" using `*`.\n",
+    "\n",
+    "```\n",
+    "SELECT\n",
+    "  TIME_PARSE(\"timestamp\") AS \"__time\",\n",
+    "  *\n",
+    "FROM \"ext\"\n",
+    "```\n",
+    "\n",
+    "The final portion of this ingestion is the `PARTITIONED BY DAY` clause which tells Driud to create a separate set of segments for each day. A `PARTITION BY` clause must be included in all `INSERT`/`REPLACE` statements."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "66b30164-b3c3-4de6-ba68-d269f0bee2d1",
+   "metadata": {},
+   "source": [
+    "#### Wait for Segment Availibility\n",
+    "In the next cell, `sql_wait_until_ready` function is used to pause until all the ingested data is available in the Historical cacheing layer before executing any queries:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "e0429f0e-ae98-472a-b686-8aa308985de6",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "druid.sql.wait_until_ready('wikipedia_events')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ca8f811a-7795-4351-bc6b-3abb34da0116",
+   "metadata": {},
+   "source": [
+    "#### Query the Data\n",
+    "Let's take a look at the data that was loaded:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "d2584e65-c952-47f1-a885-2e5d3a4fef2b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql = \"\"\"\n",
+    "SELECT channel, count(*) num_events\n",
+    "FROM \"wikipedia_events\" \n",
+    "GROUP BY 1 \n",
+    "ORDER BY 2 DESC \n",
+    "LIMIT 10\n",
```

Comment:
Suggest adding a time filter here over a LIMIT operation as it reinforces the message that Druid puts time front and center.

Reply:
done